File: mobile/lib/core/services/activity_log_service.dart
import '../constants/app_constants.dart';
import 'api_client.dart';

class ActivityLogService {
  final ApiClient _apiClient = ApiClient();

  // Get all activity logs (Admin: all logs, User: own logs)
  Future<Map<String, dynamic>> getActivityLogs({
    int page = 1,
    int limit = 20,
    String? userId,
    String? action,
    String? startDate,
    String? endDate,
  }) async {
    try {
      final queryParams = {'page': page.toString(), 'limit': limit.toString()};

      if (userId != null && userId.isNotEmpty) {
        queryParams['userId'] = userId;
      }
      if (action != null && action.isNotEmpty) {
        queryParams['action'] = action;
      }
      if (startDate != null) {
        queryParams['startDate'] = startDate;
      }
      if (endDate != null) {
        queryParams['endDate'] = endDate;
      }

      final response = await _apiClient.get(
        AppConstants.activityLogsEndpoint,
        queryParameters: queryParams,
      );

      if (response.statusCode == 200 && response.data['success'] == true) {
        return {
          'success': true,
          'logs': response.data['data']['logs'],
          'pagination': response.data['data']['pagination'],
        };
      }

      return {
        'success': false,
        'message':
            response.data['message'] ?? 'Không thể tải nhật ký hoạt động',
      };
    } catch (e) {
      return {'success': false, 'message': 'Lỗi: ${e.toString()}'};
    }
  }

}

File: frontend/src/services/activityLog.service.ts
import apiClient from './api.client';
import { API_ENDPOINTS } from '../config/api.config';

export interface ActivityLog {
  _id: string;
  userId: string | { _id: string; username?: string; email?: string; role?: string };
  username?: string;
  action: string;
  target?: string;
  resourceType?: string;
  resourceId?: string;
  resourceName?: string;
  status: 'success' | 'failed';
  timestamp?: string;
  createdAt?: string;
  updatedAt?: string;
  details?: any;
}

export interface ActivityLogParams {
  userId?: string;
  action?: string;
  status?: 'success' | 'failure';
  startDate?: string;
  endDate?: string;
  limit?: number;
  page?: number;
}

export const activityLogService = {
  async getMyLogs(params?: { limit?: number; page?: number }): Promise<ActivityLog[]> {
    const response = await apiClient.get<{ success: boolean; data: ActivityLog[] }>(
      API_ENDPOINTS.ACTIVITY_LOGS.MY_LOGS,
      { params }
    );
    return response.data.data;
  },

  async getAll(params?: ActivityLogParams): Promise<{
    success: boolean;
    data: ActivityLog[];
    pagination: {
      total: number;
      page: number;
      pages: number;
      limit: number;
    };
  }> {
    const response = await apiClient.get<{
      success: boolean;
      data: {
        logs: ActivityLog[];
        pagination: {
          total: number;
          page: number;
          pages: number;
          limit: number;
        };
      };
    }>(API_ENDPOINTS.ACTIVITY_LOGS.BASE, { params });
    return {
      success: response.data.success,
      data: response.data.data.logs || [],
      pagination: response.data.data.pagination,
    };
  },

  async getStats(params?: { startDate?: string; endDate?: string }): Promise<any> {
    const response = await apiClient.get(API_ENDPOINTS.ACTIVITY_LOGS.STATS, { params });
    return response.data;
  },
};

